src/ConnectionPool.cpp

```diff
@@ -446,8 +446,14 @@ err_code_t ConnectionPool::waitPoll() {
 
         if (pollfd_ptr->revents & (POLLERR | POLLHUP | POLLNVAL)) {
           markDeadConn(conn, keywords::kCONN_POLL_ERROR, pollfd_ptr);
-          ret_code = RET_CONN_POLL_ERR;
```

Comment:
https://github.com/douban/libmc/pull/72/files#diff-6b049cde08edceed2718a5acad05442eR427 this pr also fix some bug in ConnectionPool::waitPoll，so let us works on that pr，and merge that pr first?

Reply:
https://github.com/douban/libmc/pull/72/commits has lots of commits，but i only want to merge a few commits into master，not include releated to plog。

Reply:
Yes, I saw that PR is still WIP. Which bug it fixed? Maybe just clean it up or reopen a PR so we can work on it.

Reply:
@tclh123 that most import fix of that pr is "sys call recv should before send for best practice"。this solve erebor's "lots of noisy error" problem。yeah, i can do it before this weekend。i am asking merge it first，because i can take little care to merge your change(this pr)。i am lazy。😂

Reply:
> "sys call recv should before send for best practice"

Is that true? Actually, in `waitPoll` we need `send` first, and when first send succeed we start to POLLIN, then the order of `send` and `recv` is not important.

> this solve erebor's "lots of noisy error" problem

I think it won't fix that `recv_error` problem. When recv return 0, it means server shutdown the connection, in other words, it has been kicked.

Reply:
cc @youngsofun know some details about those commits of pr https://github.com/douban/libmc/pull/72。

Reply:
@MOON-CLJ 

Yes, it should fix. I missed 2497900 before.

It's somehow tricky but may works on most cases I think.

But,

1. what if the recv return -1 and we already send data - not working on this case. 
1. And what if the server shutdown the connection just after we sent data but not arrived yet. then we will recv 0 and the server will ignore the late request, and we reconnect but have no complete data to send.

Reply:
> can detect this half open state early

I think maybe I can combine those 2 solutions.

> and without backup iovec like rewind

Normally, it won't happen I think. It only happens, if we need to modify iovec when we send.

Reply:
@tclh123 yes，those corner cases not take in consider in pr 72。it take effort to deal with recv return 0。

> I think maybe I can combine those 2 solutions.

seems good to me。

> Normally, it won't happen I think. It only happens, if we need to modify iovec when we send.

seems has problem。if a memcached command need more than one send，the whole command need backup? send some part of the whole command in one connection ，the left part is another connection (create in reconnect)，doesnt make sense。am i right？i dont take much time to think this。

Reply:
Normally, the vector of iovec(the total send buffer) will not be modified after we dispatched. So in `rewind` we only need to re-init pointers(pointers indicate which part should be sent next).

Only if the `sendmsg` only sent partial iovec (which I think will not happen frequently), we need to modify one of the iovec, then we need to backup the vector of iovec for possibly rewind.

Reply:
https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L132 
i means if a large value set command like this 'set <key> <flags> <exptime> <10M bytes> [noreply]\r\n' ，'set <key> <flags> <exptime> <say 4M bytes>' send first，then reconnect，then sent '<left 6M bytes> [noreply]\r\n' at the another connection，these doesnt make sense from the memcached server's opinion，even if two send success，the whole command is fail。

Reply:
https://github.com/douban/libmc/blob/master/src/Connection.cpp#L234-L239 will lead to more than one send for a command。

Reply:
@MOON-CLJ 

The case you said is ok, it will resend all 10M bytes.

But you remind me that whether to resend/retry a write command is hard to determine - I don't know whether the store is succeed or failed or partial stored. And the rewind is not handle `m_requestKeys` properly. I need to reconsider this.

Reply:
@tclh123 okay, you're right。

Reply:
@tclh123 I think it's ok for `No Strong Consistency System`(like memcached). You need the client to provide a unique request id for keep idempotence even if use a strong consistency system like raft or paxos.